cli/src/commands/mod.rs

```diff
@@ -355,7 +355,7 @@ struct StatusArgs {}
 #[derive(clap::Args, Clone, Debug)]
 struct LogArgs {
     /// Which revisions to show. Defaults to the `revsets.log` setting, or
-    /// `@ | ancestors((remote_branches() | tags()).., 2)` if it is not set.
+    /// `@ | ancestors(immutable().., 2)` if it is not set.
```

Comment:
(Optional) At least until the log displays immutable commits differently, I think you should add `|immutable()` to this revset, so that the user knows which line of the log is the `main` branch. The case where this makes a difference is left as an exercise to the reader; it did happen to me.

Reply:
Actually, there are two ways this can matter, so let me show the important one:

```
A   B
*---*----* main == immutable()
     \
      * C
```

As is, `jj log` would just show `B` and `C`, and it wouldn't be entirely obvious that `B` is on `main` branch and `C` isn't.

With my suggestion, `B`, `C`, and `main` would be shown.

Reply:
I guess you excluded it because you added `tags()`. You could go with `|heads(connected(immutable()))` (if I recalled the function names correctly).

Reply:
Good point. Since `immutable()` includes ancestors (and the root commit), I don't think we need the `connected()` part. I've added `| heads(immutable())`.

Reply:
Maybe better to define `immutable_heads()` as well? `immutable()` can be large, and `x` in `x..` is evaluated eagerly at least for the default engine. Apparently `ancestors(x)..` is substituted to `x..`, but `(ancestors(x) | root())..` isn't.

Reply:
> Maybe better to define `immutable_heads()` as well?

Makes sense. Maybe we should just convert `revsets.immutable-heads` into `revset-aliases.immutable_heads()` and use that. That would mean that the expression is controlled by the user, so `root()` is not necessarily included, and `immutable_heads()` is not necessarily equal to `heads(immutable())`. I don't see much problem with that in practice. Thoughts?

Reply:
> Maybe we should just convert `revsets.immutable-heads` into `revset-aliases.immutable_heads()` and use that. That would mean that the expression is controlled by the user, so `root()` is not necessarily included, and `immutable_heads()` is not necessarily equal to `heads(immutable())`. I don't see much problem with that in practice. Thoughts?

I hope you can explain it better, because that sentence makes no sense to me.

Reply:
> Maybe we should just convert `revsets.immutable-heads` into `revset-aliases.immutable_heads()` and use that.

Suppose we're willing to merge `[revsets]` and `[revset-aliases]`, that sounds good to me. It might be also easier to define `immutable()` (a user-controllable expression) as an alias, but constructs true immutable sets programatically from `immutable_heads()` where immutability really matters.

```toml
[revset-aliases]
'immutable()' = '::immutable_heads()'
'immutable_heads()' = '...'
```

```rust
fn check_rewritable() {
    ...
    let (params, defn) = revset_aliases_map.get_function("immutable_heads");
    check params.len();
    let immutable_revset = parse_revset(defn)?.union(root()).ancestors();
    ...
```

Reply:
I'd be concerned that it would be confusing to let the user override `immutable()` and it wouldn't impact the set of commits we actually consider immutable.

I like the idea of defining it as an alias, though. Maybe we could add a config layer on top the user's config, so we override `immutable()` to always be `::(immutable_heads() | root())` internally? Or, equivalently, we could decide that user configs cannot override certain keys (maybe slightly harder to implement due to how we load revset aliases and template aliases layer by layer).

Reply:
> I hope you can explain it better, because that sentence makes no sense to me.

Yes, I'll try to remember to document how this will work in the revset documentation once we've decided.

Reply:
> I'd be concerned that it would be confusing to let the user override `immutable()` and it wouldn't impact the set of commits we actually consider immutable.

Yeah, but they can override any functions including `root()` for whatever reason, so I don't think implementing `immutable()` as a builtin function will help. We can blacklist builtin names + `immutable()` if we prefer strictness.

Reply:
> We can blacklist builtin names + immutable() if we prefer strictness.

We could have a namespace or a convention. E.g., we could say that functions starting with `__` cannot be redefined by the user and use `__immutable()` where this matters. We could either have or not have a default alias `"immutable()" = "__immutable()"`

The namespace approach would be along the lines of `buitin#immutable()`.

Reply:
> We could have a namespace or a convention. E.g., we could say that functions starting with `__` cannot be redefined by the user and use `__immutable()` where this matters.

I think the main concern is that the user can customize `immutable()` instead of `immutable_heads()` without reading the doc. Since `[revset-aliases]` is free form, it's unclear whether the alias has special meaning. Special aliases like `immutable_heads()` could be namespaced or prefixed instead, but that's basically the same as having separate `[revsets]` and `[revset-aliases]` sections.

Reply:
At the moment, do `[revset-aliases]` override `[revsets]` or vice-vera? I thought it was `[revset-aliases]`, but it sounds like we want the opposite (and that might already be the case).

Reply:
I don't think either of them override the other, but `[revset-aliases]` overrides builtins. Maybe that's what you were thinking of?

Reply:
Ah, I guess `[revsets]` currently cannot be used in `-r` arguments. (So, they *describe* revsets, but cannot be used in revsets :) ).

Reply:
I've moved the definition of `immutable()` out of this PR, so I think this is resolved.